### PR TITLE
Fix replace statements handling

### PR DESCRIPTION
**Pull request by hugares**
I was planning to submit one PR per commit to make it easier to review, https://github.com/jfrog/gocmd/pull/20 is the first one out of 3 but since I am done with all 3 changes(commits), here is a PR with all 3 commits.

Please see the individual commit messages for explanation/motivation for each changes.

Please do not squash the commits when merging this PR to preserve the information in the commit messages.

Issue https://github.com/jfrog/jfrog-cli/issues/480

### CODE REVIEW

**Comment by hugares** [comment]
@eyalbe4 Can you help getting this PR in? We are trying to upgrade to Go 1.13 and this issue is blocking us.

**Comment by eyalbe4** [comment]
Hi @hugares,

Few Questions - 

1. Why is it important to remove the replaced module from the map? In the example you shared, you are replacing a dependency with a local dependency (located inside the lib directory). Please correct me if I'm wrong, but removing it from the map only causes for this dependency not to be published to Artifdactory, if *--publish-deps* is used. If so, why wouldn't you want it get published?

2. Why is this specifically relevant to go 1.13? What changed and what is the actual problem or negative impact on the build?

Thanks!

**Comment by hugares** [comment]
> 1. Why is it important to remove the replaced module from the map? In the example you shared, you are replacing a dependency with a local dependency (located inside the lib directory). Please correct me if I'm wrong, but removing it from the map only causes for this dependency not to be published to Artifdactory, if --publish-deps is used. If so, why wouldn't you want it get published?

When a module is replaced, this means the go command will not download nor use the replaced module, it will only download and use the replacement. I changed the code to make the jfrog cli consistent with Go commands. Before my change the jfrog cli was always downloading and publishing more dependencies than needed when there are replace statements, one extra dependency per replace statement.

> 2. Why is this specifically relevant to go 1.13? What changed and what is the actual problem or negative impact on the build?

The jfrog cli uses `go mod download` to download all the dependencies from the map. Before 1.13, `go mod download` was returning 0 as exit code even when the command failed. In 1.13, they fixed that so now the jfrog cli stop downloading the dependencies as soon as it tries to download a local dependency because the `go mod download` returns 1

**Comment by eyalbe4** [comment]
Thanks for your answers @hugares.
Let me look into this again now that have better understanding of the issue.

File: executers/dependenciesutils.go [hunk 1]
```diff
@@ -396,19 +397,59 @@ func mergeReplaceDependenciesWithGraphDependencies(replaceDeps []string, graphDe
...
+		// be used as they are replaced
+		removedCount := removeDepedenciesFromGraphDependencies(graphDeps, moduleNameToReplace, moduleVersionToReplace)
+		//Only add the replacement dependency to the map if it is replacing at least one dependency
+		if removedCount > 0 {
+			// Check if the replacement dependency in the map, if not add to the map
+			_, exists := graphDeps[newDependency[0]+"@"+newDependency[1]]
+			if !exists {
+				log.Debug("Adding dependency", newDependency[0], newDependency[1])
+				graphDeps[newDependency[0]+"@"+newDependency[1]] = true
+			}
```
[the remark this answers is not in the document]
Comment (hugares) [reply executers/dependenciesutils.go]: Done

File: executers/dependenciesutils.go [hunk 2]
```diff
@@ -396,19 +397,59 @@ func mergeReplaceDependenciesWithGraphDependencies(replaceDeps []string, graphDe
...
+			removedCount = removedCount + 1
+			log.Debug("Removed", dependencyToRemove)
+		}
+	} else {
+		//otherwise remove any dependencies matching the module name, i.e any versions of the module
+		for k := range graphDeps {
+			if strings.HasPrefix(k, dependencyToRemove) {
+				delete(graphDeps, k)
+				removedCount = removedCount + 1
+				log.Debug("Removed", k)
```
[the remark this answers is not in the document]
Comment (hugares) [reply executers/dependenciesutils.go]: Done

File: executers/dependenciesutils.go [hunk 3]
```diff
@@ -396,19 +397,59 @@ func mergeReplaceDependenciesWithGraphDependencies(replaceDeps []string, graphDe
...
+
+func removeDepedenciesFromGraphDependencies(graphDeps map[string]bool, name string, version string) (removedCount int) {
+	dependencyToRemove := name + "@" + version
+	log.Debug("Dependency to remove", dependencyToRemove)
+	if version != "" {
+		//if version is specifed, remove the exact match, i.e the module with a specific version
+		if _, exists := graphDeps[dependencyToRemove]; exists {
+			delete(graphDeps, dependencyToRemove)
+			removedCount = removedCount + 1
+			log.Debug("Removed", dependencyToRemove)
```
[the remark this answers is not in the document]
Comment (hugares) [reply executers/dependenciesutils.go]: Done

File: executers/dependenciesutils.go [hunk 4]
```diff
@@ -396,19 +397,59 @@ func mergeReplaceDependenciesWithGraphDependencies(replaceDeps []string, graphDe
...
+			_, exists := graphDeps[newDependency[0]+"@"+newDependency[1]]
+			if !exists {
+				log.Debug("Adding dependency", newDependency[0], newDependency[1])
+				graphDeps[newDependency[0]+"@"+newDependency[1]] = true
+			}
+		}
+	}
+}
+
+func removeDepedenciesFromGraphDependencies(graphDeps map[string]bool, name string, version string) (removedCount int) {
```
[the remark this answers is not in the document]
Comment (hugares) [reply executers/dependenciesutils.go]: Done

File: executers/dependenciesutils.go [hunk 5]
```diff
@@ -396,19 +397,59 @@ func mergeReplaceDependenciesWithGraphDependencies(replaceDeps []string, graphDe
...
+	dependencyToRemove := name + "@" + version
+	log.Debug("Dependency to remove", dependencyToRemove)
+	if version != "" {
+		//if version is specifed, remove the exact match, i.e the module with a specific version
```
[the remark this answers is not in the document]
Comment (hugares) [reply executers/dependenciesutils.go]: Done

File: executers/dependenciesutils.go [hunk 6]
```diff
@@ -396,19 +397,59 @@ func mergeReplaceDependenciesWithGraphDependencies(replaceDeps []string, graphDe
...
+		if _, exists := graphDeps[dependencyToRemove]; exists {
+			delete(graphDeps, dependencyToRemove)
+			removedCount = removedCount + 1
```
[the remark this answers is not in the document]
Comment (hugares) [reply executers/dependenciesutils.go]: I did as you suggested and then I got that message from go-lint:
`[go-lint] should replace removedCount += 1 with removedCount++`

So went for the suggestion from go-lint which is even shorter.

**Comment by hugares** [comment]
Thanks for the review!

**Comment by bguendy** [comment]
@eyalbe4 can you please help to get this PR in ?

**Comment by eyalbe4** [comment]
Thank you so much for @bguendy!
We will issue a JFrog CLI release soon.

**Comment by hugares** [comment]
@eyalbe4 thanks for merging, can you tag this repo (v0.1.11?) so we can update jfrog-cli with a released version?

**Comment by eyalbe4** [comment]
Sure @hugares. **v0.1.11** is released now.

**Comment by bguendy** [comment]
@eyalbe4 Thanks. Here is the PR to upgrade jfrog-cli to this version:
https://github.com/jfrog/jfrog-cli/pull/493

**Pull request merged**